internal/app/prometheus-multi-tenant-proxy/jwt.go

```diff
@@ -0,0 +1,182 @@
+	}
+	s += "}"
+	return s
+}
+
+// Load loads or reloads the JWKS from its config location (file or URL).
+func (auth *JwtAuth) Load() bool {
+	if auth.config == "" {
+		log.Fatalf("JWTAuth: Load() cannot be called without a config")
+	}
+
+	if auth.isFile {
+		return auth.loadFromFile(&auth.config)
+	}
+	return auth.loadFromURL(&auth.config)
+
+}
+
+func (auth *JwtAuth) loadFromURL(url *string) bool {
+	// We do not use the jwks.Reload() method here
+	// to avoid getting the lock unless strictly necessary.
+	jwks, err := keyfunc.Get(*url, keyfunc.Options{})
+	if err != nil {
+		log.Printf("Failed to get the JWKS from the given URL: %s", err)
+		return false
+	}
+	b64content := base64.StdEncoding.EncodeToString(jwks.RawJWKS())
+
+	if auth.jwks == nil || b64content != auth.b64content {
+		auth.lock.RLock()
+		defer auth.lock.RUnlock()
+		auth.jwks = jwks
+		auth.b64content = b64content
+		log.Printf("Reloaded JWKS from URL: %s", *url)
+	}
+	return true
+}
+
+func (auth *JwtAuth) loadFromFile(location *string) bool {
+	content, err := ioutil.ReadFile(*location)
+	if err != nil {
+		log.Printf("Failed to read JWKS file: %v", err)
+		return false
+	}
+	b64content := base64.StdEncoding.EncodeToString(content)
+	if auth.b64content == b64content {
+		// nothing to do
+		return true
+	}
+
+	jwks, err := keyfunc.NewJSON(json.RawMessage(content))
+	if err != nil {
+		log.Printf("Failed to parse JWKS file: %v", err)
+		return false
+	}
+	auth.lock.RLock()
+	defer auth.lock.RUnlock()
+	auth.b64content = b64content
+	auth.jwks = jwks
+	log.Print("Reloaded JWKS from file")
+	return true
+}
+
+// IsAuthorized validates the user by verifying the JWT token in
+// the request and returning the namespaces claim found in token the payload.
+func (auth *JwtAuth) IsAuthorized(r *http.Request) (bool, []string) {
+	tokenString := extractTokens(&r.Header)
+	if tokenString == "" {
+		log.Printf("Token is missing from header request")
+		return false, nil
+	}
+	return auth.isAuthorized(tokenString)
+}
+
+// WriteUnauthorisedResponse writes a 401 Unauthorized HTTP response
+func (auth *JwtAuth) WriteUnauthorisedResponse(w http.ResponseWriter) {
+	w.WriteHeader(401)
+	w.Write([]byte("Unauthorised\n"))
+}
+
+func (auth *JwtAuth) isAuthorized(tokenString string) (bool, []string) {
+	token, err := jwt.ParseWithClaims(tokenString, &NamespaceClaim{}, auth.jwks.Keyfunc)
+	if err != nil || !token.Valid {
+		log.Printf("%s\n", err)
+		return false, nil
+	}
+
+	claims := token.Claims.(*NamespaceClaim)
+	if len(claims.Namespaces) == 0 {
+		log.Printf("token claim is invalid: namespaces is missing or empty")
+		return false, nil
+	}
+	return true, claims.Namespaces
+}
+
+func isValidSigningMethod(signingMethod string) bool {
+	for _, alg := range jwt.GetAlgorithms() {
+		if signingMethod == alg {
+			return true
+		}
+	}
+	return false
+}
+
+func extractTokens(headers *http.Header) string {
+	if token := headers.Get("Authorization"); token != "" {
```

Comment:
Do you think it could make sense adding information in the `README.md` file about the header priority?
AFAIK, 
- It tries `Authorization` header first
- If `Authorization` header is not present, it tries the `token` one.